src/rust/plugin-work-queue/README.md

```diff
@@ -0,0 +1,106 @@
+$$
+BEGIN
+   RETURN bytes * 1000 * 1000;
+END;
+$$ LANGUAGE plpgsql;
+
+CREATE TABLE IF NOT EXISTS plugin_work_queue.generator_plugin_executions
+(
+   execution_key    bigserial                  NOT NULL,
+   tenant_id        uuid                       NOT NULL,
+   plugin_id        uuid                       NOT NULL,
+   pipeline_message bytea                      NOT NULL,
+   status           plugin_work_queue.status_t NOT NULL,
+   creation_time    timestamptz                NOT NULL,
+   last_updated     timestamptz                NOT NULL,
+   visible_after    timestamptz,
+   try_count        integer                    NOT NULL,
+   CHECK (length(pipeline_message) < plugin_work_queue.megabytes(256)),
+   CHECK (last_updated >= creation_time)
+)
+   PARTITION BY RANGE (creation_time);
+
+CREATE TABLE IF NOT EXISTS plugin_work_queue.analyzer_plugin_executions
+(
+   execution_key    bigserial                  NOT NULL,
+   tenant_id        uuid                       NOT NULL,
+   plugin_id        uuid                       NOT NULL,
+   pipeline_message bytea                      NOT NULL,
+   status           plugin_work_queue.status_t NOT NULL,
+   creation_time    timestamptz                NOT NULL,
+   last_updated     timestamptz                NOT NULL,
+   visible_after    timestamptz,
+   try_count        integer                    NOT NULL,
+   CHECK (length(pipeline_message) < plugin_work_queue.megabytes(256)),
+   CHECK (last_updated >= creation_time)
+)
+   PARTITION BY RANGE (creation_time);
+
+CREATE INDEX IF NOT EXISTS execution_key_ix ON plugin_work_queue.generator_plugin_executions (execution_key);
+CREATE INDEX IF NOT EXISTS creation_time_ix ON plugin_work_queue.generator_plugin_executions (creation_time);
+
+CREATE INDEX IF NOT EXISTS execution_key_ix ON plugin_work_queue.analyzer_plugin_executions (execution_key);
+CREATE INDEX IF NOT EXISTS creation_time_ix ON plugin_work_queue.analyzer_plugin_executions (creation_time);
+```
+
+`execution_key` is a `bigserial` - an auto-incrementing 64bit integer.
+
+`plugin_id` is the uuid name of the plugin to send this execution result to.
+
+`pipeline_message` is the raw bytes to be interpreted by the plugin. The message is limited to 256MB as an arbitrary but
+reasonable upper limit.
+
+`status` is an enum representing the state. 'enqueued' is the default, and a message will be set to either 'processed' or 'failed' based on its success.
+
+`creation_time` is when the row was created.
+
+`last_updated` is set with each update to the row
+
+`visible_after` is essentially a visibility timeout. A `visible_after` of NULL means a message is available, which
+is the default state. Every time a message is SELECT'd we update `visible_after`. See the `Visibility Timeout` section
+below.
+
+`try_count` on every receive we increment `try_count` to indicate another attempt to process this message
+
+
+## Algorithm
```

Comment:
After reading through this algorithm, I think a few changes would help to make it significantly clearer.

I think the various concepts the algorithm mentions need to be defined first, and should be defined in generic terms. For instance, defining "visible" as "has a timestamp that is less than the current timestamp" is true in the strictest mechanical sense, but doesn't say anything about _why_ this is something that is important.

Some important concepts I can see that need description:
- age and "aging out". Why do we care about age, why is 1 day the limit that is chosen, what are the implications for messages that are too old?
- the set of statuses a message may have, and what governs the transitions between statuses. In particular, it seems that we have an implicit status of "dispatched" that is not part of the SQL status enum, but is currently implemented as "being invisible". Not being part of the enum is fine if that's not how we need to implement it, but it should be present as a concept (and our algorithm should be expressed in terms of it). (The implementation could also be reworked so this is more readily apparent in the code, even if it's not part of the enum; it could be incorporated into a view, for instance. Actually, I think views could help to clarify _a lot_ of this logic. For instance, a view could be based on both the status enum and the visibility timeout, providing a unified representation of the _conceptual_ state. Our status enum would thus be a _partial_ representation of the complete state space, and would simply be part of the implementation details.)
- the notion of "visibility". As written, the idea is spread throughout three separate parts of this file (the schema, the algorithm, and the dedicated "Visibility Timeout" section). It would be much better to consolidate all this information into a single description, along with the rest of the core concepts. The value of the interval being added to the visibility timestamp is also not described, nor the rationale for choosing its static value.
- retries. Under what conditions will a message be retried? Can a message end up being dispatched to multiple workers at the same time? What are the implications of that? Is there a maximum amount of times a message may be retried? What happens if that amount is exceeded?

Reply:
I've attempted to clarify some of this in the readme.